### PR TITLE
refactor($location): move repeated path normalization code into helper method

### PR DESCRIPTION
**Pull request by jbedard**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Refactor to reduce duplicate code

**Other information**:
I originally did this as part of an attempted solution in #16611, but I thought it might be worth committing on it's own.

### CODE REVIEW

File: src/ng/location.js [hunk 1]
```diff
@@ -143,18 +151,8 @@ function LocationHtml5Url(appBase, appBaseNoFile, basePrefix) {
...
-   */
-  this.$$compose = function() {
-    var search = toKeyValue(this.$$search),
-        hash = this.$$hash ? '#' + encodeUriSegment(this.$$hash) : '';
-
-    this.$$url = encodePath(this.$$path) + (search ? '?' + search : '') + hash;
-    this.$$absUrl = appBaseNoFile + this.$$url.substr(1); // first char is always '/'
-
-    this.$$urlUpdatedByLocation = true;
+  this.$$normalizeUrl = function(path) {
```
Comment (gkalpak) [inline src/ng/location.js]: I think `$$normalizeUrl`'s argument should be named `url` (not `path`), because that is what it is (in `$location` terms).

**Review by gkalpak — APPROVED** [review]
One minor suggestion. Otherwise LGTM.

File: src/ng/location.js (hunk 1 above)
Comment (jbedard) [reply src/ng/location.js]: Hmm... even though it's actually (part of) a path? The main thing this method does is append a base path to the argument. Seems like this is more of a path2url method...

Comment (gkalpak) [reply src/ng/location.js]: It is not necessarilt part of the path. It also contains query and hash fragment. So, technically it is part the browser URL.
But in `$location` terms, it is the URL (which is why `$$normalizeUrl()` is called with `$location.$$url` :grin:).

**Comment by jbedard** [comment]
Updated

**Pull request merged**